### PR TITLE
C3XGate(angle=...) deprecation

### PR DESCRIPTION
**Pull request by 1ucian0**
### Summary

The `C3XGate` class supports an `angle` parameter mostly because the need of a `C3SqrtXGate`. This turns `C3XGate` into a generic `C3RXGate` (which support `angle`) and makes `C3SqrtXGate` and `C3XGate` subclasses of `C3RXGate`.

### CODE REVIEW

File: qiskit/circuit/library/standard_gates/x.py [hunk 1]
```diff
@@ -570,6 +569,27 @@ def inverse(self):
...
+        elif angle == numpy.pi / 4:
+            warnings.warn('The parameter angle in C3XGate is going to be removed in the future. '
+                          'angle=pi/8, while be the default behavior C3XGate.',
+                          DeprecationWarning, stacklevel=2)
+        elif angle is not None:
+            warnings.warn('The parameter angle in C3XGate is going to be removed in the future. '
+                          'When angle!=pi/8, use C3RXGate()',
+                          DeprecationWarning, stacklevel=2)
+
+        super().__init__(angle=angle or numpy.pi/4, label=label, ctrl_state=ctrl_state)
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/x.py:588]: Does this work if angle is 0? Maybe we need a proper `is None` comparison

Comment (1ucian0) [reply qiskit/circuit/library/standard_gates/x.py]: good catch. Fixed in a3800b8ca

File: qiskit/circuit/library/standard_gates/x.py [hunk 2]
```diff
@@ -548,28 +565,50 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
...
+class C3XGate(C3RXGate):
+    def __init__(self, angle=None, label=None, ctrl_state=None):
+        import warnings
+        if angle == numpy.pi / 8:
+            warnings.warn('The parameter angle in C3XGate is going to be removed in the future. '
+                          'When angle=pi/8, use C3SqrtXGate()',
+                          DeprecationWarning, stacklevel=2)
+        elif angle == numpy.pi / 4:
+            warnings.warn('The parameter angle in C3XGate is going to be removed in the future. '
+                          'angle=pi/8, while be the default behavior C3XGate.',
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/x.py:579]:
```suggestion
                          'angle=pi/4, while be the default behavior C3XGate.',
```

File: qiskit/circuit/library/standard_gates/x.py [hunk 3]
```diff
@@ -548,28 +565,50 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
...
+                          DeprecationWarning, stacklevel=2)
+        elif angle is not None:
+            warnings.warn('The parameter angle in C3XGate is going to be removed in the future. '
+                          'When angle!=pi/8, use C3RXGate()',
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/x.py:583]:
```suggestion
                          'When angle!=pi/8 or pi/4, use C3RXGate()',
```

Comment (1ucian0) [reply qiskit/circuit/library/standard_gates/x.py]: I fixed it in 844228edddaa8f5b127994e60f45ccf3b2fcab56

**Comment by 1ucian0** [comment]
`on hold` until #4638 is merged.

File: qiskit/circuit/library/standard_gates/x.py [hunk 4]
```diff
@@ -539,36 +532,68 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
...
+            h d; cu1(-pi/8) c,d; h d;
+            cx b,c;
+            h d; cu1(pi/8) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/8) c,d; h d;
+        }
+        """
+        from .u1 import CU1Gate
+
+        controlled_v = Instruction('cv', 2, 0, [])
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/x.py:588]: It should, waiting for it to be merged 👍

File: qiskit/circuit/library/standard_gates/x.py [hunk 5]
```diff
@@ -539,36 +532,68 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
...
+    def __init__(self, label=None, ctrl_state=None):
+        """The 3-qubit controlled √X gate.
+
+        Args:
+            label (str or None): An optional label for the gate [Default: None]
+            ctrl_state (int or str or None): control state expressed as integer,
+                string (e.g. '110'), or None. If None, use all 1s.
+        """
+        super().__init__('c3sqrtx', 4, [], label=label, num_ctrl_qubits=3, ctrl_state=ctrl_state)
+        self.base_gate = Gate('sqrtx', 1, [])
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/x.py:564]: If we have the SXGate base class it depends. We need to check if this decomposition is more efficient than the default mechanism to construct the 3-controlled version, also blocked by #4638.

[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/x.py:564]: With #4638 and `SXGate().control(3)`:
```python
>>> transpiled = transpile(circuit, basis_gates=['u1','u2','u3','cx'], optimization_level=0)
>>> transpiled.count_ops()
OrderedDict([('cx', 120), ('u1', 112), ('u3', 30), ('u2', 4)])
>>> transpiled = transpile(circuit, basis_gates=['u1','u2','u3','cx'], optimization_level=3)
>>> transpiled.count_ops()
OrderedDict([('cx', 100), ('u1', 98), ('u3', 2), ('u2', 1)])
```
the decomposition in this PR
```python
>>> transpiled = transpile(circuit, basis_gates=['u1','u2','u3','cx'], optimization_level=0)
>>> transpiled.count_ops()
OrderedDict([('u1', 21), ('cx', 20), ('u2', 14)])
>>> transpiled = transpile(circuit, basis_gates=['u1','u2','u3','cx'], optimization_level=3)
>>> transpiled.count_ops()
OrderedDict([('u1', 20), ('cx', 20), ('u2', 2)])
```
Looking at the CX count I think we should add this class.

**Comment by 1ucian0** [comment]
#4638 merged. removing `on hold`

File: qiskit/circuit/library/standard_gates/x.py [hunk 6]
```diff
@@ -539,36 +532,68 @@ def control(self, num_ctrl_qubits=1, label=None, ctrl_state=None):
...
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
-    #                         [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=complex)
+    def to_matrix(self):
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/x.py:545]: This is correct only if `angle==pi/4`, right?

Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/x.py:545]: Yes. In principle we could compute the matrix for any `angle` and return that

**Comment by 1ucian0** [comment]
PR #5668 is already taking care of this. Closing.

**Pull request closed without merging**